File: reticle-core/src/main/kotlin/dev/reticle/core/Protocol.kt
package dev.reticle.core

import kotlinx.serialization.Serializable

/**
 * Wire protocol shared by the in-app server and the host CLI: a small set of
 * localhost endpoints (/runtime, /report, /snapshot, /semantics, /logs).
 *
 * Endpoint map (HTTP GET unless noted), all bound to 127.0.0.1 in-app and
 * reached through `adb forward`:
 *
 *   GET  /runtime         -> RuntimeInfo
 *   GET  /report          -> UiReport
 *   GET  /snapshot        -> Snapshot
 *   GET  /semantics       -> SemanticTree
 *   GET  /compact         -> CompactObservation
 *   GET  /logs            -> LogBatch
 *   GET  /screenshot      -> image/png bytes
 *   GET  /keyboard        -> KeyboardInfo
 *   POST /keyboard/hide   -> KeyboardHideResult (no body)
 *   POST /editor-action   -> EditorActionResult (no body)
 *   POST /mutate          -> MutationResult   (body: MutationRequest)
 *   POST /clipboard       -> "ok"             (body: raw UTF-8 text)
 *
 * The Swift `Endpoints` in reticle-swift mirrors this list, with two
 * deliberate asymmetries where the platform affordance exists on one side
 * only:
 *
 * - `/editor-action` ([EDITOR_ACTION]) is Android-only; UIKit has no
 *   equivalent "perform the return key's action" entry point, so `act type
 *   --submit` on iOS goes through the HID return key.
 * - `/activate` is iOS-only (`Endpoints.activate` on the Swift side, absent
 *   here): it exists because host-side HID synthesis cannot reach a real iOS
 *   device, which is not a problem Android has.
 *
 * Anything else in either list without a twin is drift, not design.
 */
object Endpoints {
    const val RUNTIME = "/runtime"
    const val REPORT = "/report"
    const val SNAPSHOT = "/snapshot"
    const val SEMANTICS = "/semantics"
    const val COMPACT = "/compact"
    const val LOGS = "/logs"
    const val SCREENSHOT = "/screenshot"
    const val MUTATE = "/mutate"

    /**
     * Set the device clipboard from inside the app process (body is the raw
     * UTF-8 text). This is the reliable way to stage non-ASCII input: `adb shell
     * input text` cannot emit non-ASCII at all, host-side `adb shell cmd
     * clipboard` is absent on many OEM builds, and Android 10+ blocks clipboard
     * writes from any process that isn't the foreground app — which the agent,
     * running inside that app, is. The CLI follows this with a KEYCODE_PASTE.
     */
    const val CLIPBOARD = "/clipboard"

    /** Current system-keyboard (IME) state, probed from inside the app. */
    const val KEYBOARD = "/keyboard"

    /**
     * Dismiss the system keyboard from inside the app process via
     * InputMethodManager — deterministic, unlike a host-side KEYCODE_BACK
     * (which navigates back when the keyboard is already gone). Answers with
     * the settled post-hide state.
     */
    const val KEYBOARD_HIDE = "/keyboard/hide"

    /**
     * Android-only (no iOS counterpart — see the object doc).
     *
     * Perform the focused field's IME editor action (the keyboard's Done /
     * Next / Go / Search / Send key) from inside the app process. This is what
     * `act type --submit` prefers: TextView.onEditorAction() drives the app's
     * OnEditorActionListener (React Native's onSubmitEditing included)
     * deterministically, where a host-side KEYCODE_ENTER inserts a newline
     * into multiline fields and is dropped by some IMEs. Answers with the
     * settled post-action keyboard state.
     */
    const val EDITOR_ACTION = "/editor-action"
}

/** Identifies the running app process behind the loopback server. */
@Serializable
data class RuntimeInfo(
    val packageName: String,
    val processName: String,
    val pid: Int,
    val sdkInt: Int,
    val agentVersion: String,
    val port: Int,
)

@Serializable
data class LogEntry(
    val timestampMillis: Long,
    val level: String,
    val message: String,
    val metadata: Map<String, MetadataValue> = emptyMap(),
)

@Serializable
data class LogBatch(val entries: List<LogEntry>)

/**
 * A single-capture UI report produced inside the app process.
 *
 * The full snapshot is captured once, then the semantic tree and compact
 * observation are derived from that exact frame so report files cannot drift.
 */
@Serializable
data class UiReport(
    val snapshot: Snapshot,
    val semantics: SemanticTree,
    val compact: CompactObservation,
) {
    companion object {
        /** Build all report views from one authoritative snapshot. */
        fun from(snapshot: Snapshot): UiReport = UiReport(
            snapshot = snapshot,
            semantics = SemanticTree.build(snapshot),
            compact = CompactObservation.from(snapshot),
        )
    }
}

/** Answer of [Endpoints.KEYBOARD_HIDE]: what was on screen, and what is now. */
@Serializable
data class KeyboardHideResult(
    val wasVisible: Boolean,
    val keyboard: KeyboardInfo,
)

/**
 * Answer of [Endpoints.EDITOR_ACTION]: whether a focused text field performed
 * its IME action, which action it was ("done"/"next"/"go"/"search"/"send"),
 * and the settled keyboard state afterwards.
 */
@Serializable
data class EditorActionResult(
    val performed: Boolean,
    val action: String? = null,
    val keyboard: KeyboardInfo? = null,
    val message: String? = null,
)

/**
 * Runtime property mutation. Allowlisted: only a bounded set of View properties
 * may be patched live so UI diagnosis and design iteration can happen without
 * rebuilding.
 */
@Serializable
data class MutationRequest(
    val selector: Selector,
    val property: String, // e.g. "alpha", "visibility", "text", "backgroundColor"
    val value: MetadataValue,
)

@Serializable
data class MutationResult(
    val applied: Boolean,
    val ref: String? = null,
    val previousValue: MetadataValue? = null,
    val message: String? = null,
)

/**
 * A stable target for actions and mutations. Resolution order: testId /
 * resource-id first (semantic-tree-backed), then ref, then raw point.
 */
@Serializable
data class Selector(
    val testId: String? = null,
    val resourceId: String? = null,
    /** CSS selector for DOM nodes captured from an embedded WebView. */
    val cssSelector: String? = null,
    val ref: String? = null,
    val point: Point? = null,
    /**
     * Visible text or a11y label, for controls the framework builds without any
     * stable id: a `Spinner` dropdown's rows, a `PopupMenu`'s items, an alert's
     * buttons. Those are captured but carry only a SHARED resource id (`text1`,
     * `title`), so nothing else can single one out. Matching is exact first, then
     * substring, and AMBIGUITY IS AN ERROR — never a silent first-match, which is
     * the failure mode that makes a wrong tap look like a working one.
     */
    val label: String? = null,
    /**
     * A substring/region within a node. Combined with a node selector
     * (testId/resourceId/ref) it targets a span/virtual-node/char-range inside
     * that node — e.g. tap just the 《agreement》 link inside a checkbox row.
     */
    val region: String? = null,
) {
    fun describe(): String = buildString {
        append(
            testId?.let { "testId=$it" }
                ?: resourceId?.let { "resourceId=$it" }
                ?: cssSelector?.let { "css=$it" }
                ?: ref?.let { "ref=$it" }
                ?: point?.let { "point=${it.x},${it.y}" }
                ?: label?.let { "label=$it" }
                ?: "<empty>"
        )
        region?.let { append(" region=\"$it\"") }
    }
}

File: reticle-swift/Sources/ReticleProtocol/Protocol.swift
import Foundation

/// Wire endpoints shared by the in-app server and the host. Mirrors
/// reticle-core's `Endpoints` — with two deliberate asymmetries, because the
/// underlying platform affordance exists on one side only:
///
/// - `activate` (`/activate`) is **iOS-only**. It has no Android counterpart:
///   there, host-side HID synthesis reaches a real device, so an in-process
///   activation path is not needed.
/// - `typeText` (`/type`) is **iOS-only**, for the same reason as `activate`:
///   a real device has no host-reachable HID keyboard, so the text is inserted
///   from inside the app through UIKit's own text-input pipeline.
/// - `/editor-action` is **Android-only** (`Endpoints.EDITOR_ACTION` in
///   reticle-core, absent here). It drives `TextView.onEditorAction()`; the iOS
///   analogue is not a separate endpoint but `typeText`'s `submit` flag, which
///   fires the return key's action in-process (`textFieldShouldReturn` +
///   `.editingDidEndOnExit`). On a simulator `act type --submit` still goes
///   through the HID return key.
///
/// Anything else in either list without a twin is drift, not design.
public enum Endpoints {
    public static let runtime = "/runtime"
    public static let report = "/report"
    public static let snapshot = "/snapshot"
    public static let semantics = "/semantics"
    public static let compact = "/compact"
    public static let logs = "/logs"
    public static let screenshot = "/screenshot"
    public static let mutate = "/mutate"
    public static let clipboard = "/clipboard"

    /// iOS-only (no Android counterpart — see the type doc).
    /// In-process control activation (POST, body: ActivationRequest). The agent
    /// resolves the selector to a control and fires its action from *inside* the
    /// app, so it works on a real device where host-side HID synthesis cannot
    /// reach — the on-device analogue of a tap. Limited to activatable targets.
    public static let activate = "/activate"

    /// iOS-only (no Android counterpart — see the type doc).
    /// In-process text entry (POST, body: TypeTextRequest). The agent focuses the
    /// resolved field and inserts the text through `UIKeyInput.insertText` — the
    /// same entry point the system keyboard uses, so delegates, `.editingChanged`
    /// and SwiftUI bindings all fire. This is the on-device analogue of `act type`,
    /// where host-side HID key synthesis cannot reach.
    public static let typeText = "/type"

    /// Current system-keyboard state, probed from inside the app (GET).
    public static let keyboard = "/keyboard"

    /// Dismiss the system keyboard from inside the app process (POST, no body):
    /// resignFirstResponder on iOS, InputMethodManager on Android. Answers with
    /// the settled post-hide state.
    public static let keyboardHide = "/keyboard/hide"
}

/// Answer of `Endpoints.keyboardHide`: what was on screen, and what is now.
/// Mirrors reticle-core's `KeyboardHideResult`.
public struct KeyboardHideResult: Codable, Sendable {
    public var wasVisible: Bool
    public var keyboard: KeyboardInfo

    public init(wasVisible: Bool, keyboard: KeyboardInfo) {
        self.wasVisible = wasVisible
        self.keyboard = keyboard
    }
}

/// Request to activate a control in-process (the on-device "tap").
public struct ActivationRequest: Codable, Sendable {
    public var selector: Selector
    public init(selector: Selector) { self.selector = selector }
}

/// Three-state, for the same reason `WaitVerdict` is: the middle state is the
/// point. Collapsing it into `refused` is how an observer lies — measured on an
/// iPhone 13 Pro Max / iOS 26, `UITextView.accessibilityActivate()` OPENS the
/// link it contains (the delegate's `shouldInteractWith` runs) and returns
/// `false` anyway, so a reader that trusts the Bool reports "nothing happened"
/// about a screen that has already navigated.
public enum ActivationOutcome: String, Codable, Sendable {
    /// Dispatched, and the target said so.
    case activated
    /// Dispatched, and the target answered `false` — which UIKit also does for
    /// activations it PERFORMED. Nothing in-process can tell the two apart, so
    /// the caller has to look: `--verify` / `--trace-output` (which run for this
    /// outcome) or a fresh observation.
    case unconfirmed
    /// Nothing was dispatched — no node matched, or the resolved target has no
    /// activation surface at all (a text-range region, a plain view).
    case refused
}

/// Result of an in-process activation. `activated` is true only for
/// `outcome == .activated`; see `ActivationOutcome` for why `false` is two
/// different answers and why they must not be merged.
public struct ActivationResult: Codable, Sendable {
    public var activated: Bool
    public var ref: String?
    public var typeName: String?
    public var via: String?
    public var message: String?
    /// Absent from an agent that predates the three-state answer; read it
    /// through `resolvedOutcome`, which falls back to the old two-state meaning.
    public var outcome: ActivationOutcome?

    public init(activated: Bool, ref: String? = nil, typeName: String? = nil, via: String? = nil,
                message: String? = nil, outcome: ActivationOutcome? = nil) {
        self.activated = activated
        self.ref = ref
        self.typeName = typeName
        self.via = via
        self.message = message
        self.outcome = outcome
    }

    /// The outcome, with the pre-three-state fallback applied: an older agent
    /// reporting `activated=false` meant "refused", which is what it said then.
    public var resolvedOutcome: ActivationOutcome {
        outcome ?? (activated ? .activated : .refused)
    }
}

/// Request to type into a field from inside the app (the on-device "type").
///
/// `selector` is optional: with none, the text goes to whatever holds focus —
/// the same rule the HID path follows. The host resolves a `--label` / region
/// selector to a `ref` before sending, so the agent only ever matches by
/// `testId` / `ref` (its own resolver has no label channel).
public struct TypeTextRequest: Codable, Sendable {
    public var selector: Selector?
    public var text: String
    /// Empty the field before typing, and prove it is empty in the result.
    public var clear: Bool
    /// Fire the return key's action after the text lands.
    public var submit: Bool
    /// Gap between characters, for a field whose formatter drops keystrokes it
    /// receives in one burst (`act type --type-delay`).
    public var perCharDelayMs: Int?

    public init(selector: Selector? = nil, text: String, clear: Bool = false,
                submit: Bool = false, perCharDelayMs: Int? = nil) {
        self.selector = selector
        self.text = text
        self.clear = clear
        self.submit = submit
        self.perCharDelayMs = perCharDelayMs
    }
}

/// What `--clear` did in-process, and whether the field is provably empty.
/// The iOS agent's twin of the host-side `ClearOutcome` on the HID path.
public struct TypeClearOutcome: Codable, Sendable {
    public var emptied: Bool
    public var before: String?
    public var after: String?
    public var deletes: Int
    public var unavailable: String?

    public init(emptied: Bool, before: String? = nil, after: String? = nil,
                deletes: Int = 0, unavailable: String? = nil) {
        self.emptied = emptied
        self.before = before
        self.after = after
        self.deletes = deletes
        self.unavailable = unavailable
    }
}

/// Result of in-process typing. `typed` is false with a `message` when no
/// typeable field could be resolved or focused — Reticle never reports text it
/// did not land. `before`/`after` are the field read back around the insert
/// (masked for a secure field, as everywhere else).
public struct TypeTextResult: Codable, Sendable {
    public var typed: Bool
    public var ref: String?
    public var typeName: String?
    public var via: String?
    public var before: String?
    public var after: String?
    public var secure: Bool?
    public var cleared: TypeClearOutcome?
    /// How the return key's action was fired, when `submit` was asked for.
    public var submitted: String?
    public var message: String?

    public init(typed: Bool, ref: String? = nil, typeName: String? = nil, via: String? = nil,
                before: String? = nil, after: String? = nil, secure: Bool? = nil,
                cleared: TypeClearOutcome? = nil, submitted: String? = nil, message: String? = nil) {
        self.typed = typed
        self.ref = ref
        self.typeName = typeName
        self.via = via
        self.before = before
        self.after = after
        self.secure = secure
        self.cleared = cleared
        self.submitted = submitted
        self.message = message
    }
}

/// Identifies the running app process behind the loopback server. The Android
/// field names are kept verbatim for wire compatibility; on iOS `packageName`
/// carries the bundle identifier and `sdkInt` the major OS version.
public struct RuntimeInfo: Codable, Sendable {
    public var packageName: String
    public var processName: String
    public var pid: Int
    public var sdkInt: Int
    public var agentVersion: String
    public var port: Int

    public init(packageName: String, processName: String, pid: Int, sdkInt: Int, agentVersion: String, port: Int) {
        self.packageName = packageName
        self.processName = processName
        self.pid = pid
        self.sdkInt = sdkInt
        self.agentVersion = agentVersion
        self.port = port
    }
}

public struct LogEntry: Codable, Sendable {
    public var timestampMillis: Int64
    public var level: String
    public var message: String
    public var metadata: [String: MetadataValue]

    public init(timestampMillis: Int64, level: String, message: String, metadata: [String: MetadataValue] = [:]) {
        self.timestampMillis = timestampMillis
        self.level = level
        self.message = message
        self.metadata = metadata
    }

    private enum CodingKeys: String, CodingKey { case timestampMillis, level, message, metadata }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(timestampMillis, forKey: .timestampMillis)
        try c.encode(level, forKey: .level)
        try c.encode(message, forKey: .message)
        if !metadata.isEmpty { try c.encode(metadata, forKey: .metadata) }
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        timestampMillis = try c.decode(Int64.self, forKey: .timestampMillis)
        level = try c.decode(String.self, forKey: .level)
        message = try c.decode(String.self, forKey: .message)
        metadata = try c.decodeIfPresent([String: MetadataValue].self, forKey: .metadata) ?? [:]
    }
}

public struct LogBatch: Codable, Sendable {
    public var entries: [LogEntry]
    public init(entries: [LogEntry]) { self.entries = entries }
}

/// A single-capture UI report produced inside the app process: one snapshot,
/// with the semantic tree and compact observation derived from that exact frame.
public struct UiReport: Codable, Sendable {
    public var snapshot: Snapshot
    public var semantics: SemanticTree
    public var compact: CompactObservation

    public init(snapshot: Snapshot, semantics: SemanticTree, compact: CompactObservation) {
        self.snapshot = snapshot
        self.semantics = semantics
        self.compact = compact
    }

    /// Build all report views from one authoritative snapshot.
    public static func from(_ snapshot: Snapshot) -> UiReport {
        UiReport(
            snapshot: snapshot,
            semantics: SemanticTree.build(from: snapshot),
            compact: CompactObservation.from(snapshot)
        )
    }
}

/// A stable target for actions and mutations. Resolution order: testId /
/// resourceId first, then ref, then raw point.
public struct Selector: Codable, Sendable {
    public var testId: String?
    public var resourceId: String?
    public var cssSelector: String?
    public var ref: String?
    public var point: Point?
    /// Visible text or a11y label, for framework controls with no stable id
    /// (Spinner rows, PopupMenu items, alert buttons). Exact match first, then
    /// substring; ambiguity is an error, never a silent first match.
    public var label: String?
    public var region: String?

    public init(testId: String? = nil, resourceId: String? = nil, cssSelector: String? = nil, ref: String? = nil, point: Point? = nil, label: String? = nil, region: String? = nil) {
        self.testId = testId
        self.resourceId = resourceId
        self.cssSelector = cssSelector
        self.ref = ref
        self.point = point
        self.label = label
        self.region = region
    }

    public func describe() -> String {
        var base = "<empty>"
        if let testId { base = "testId=\(testId)" }
        else if let resourceId { base = "resourceId=\(resourceId)" }
        else if let cssSelector { base = "css=\(cssSelector)" }
        else if let ref { base = "ref=\(ref)" }
        else if let point { base = "point=\(point.x),\(point.y)" }
        else if let label { base = "label=\(label)" }
        if let region { return "\(base) region=\"\(region)\"" }
        return base
    }
}

public struct MutationRequest: Codable, Sendable {
    public var selector: Selector
    public var property: String
    public var value: MetadataValue

    public init(selector: Selector, property: String, value: MetadataValue) {
        self.selector = selector
        self.property = property
        self.value = value
    }
}

public struct MutationResult: Codable, Sendable {
    public var applied: Bool
    public var ref: String?
    public var previousValue: MetadataValue?
    public var message: String?

    public init(applied: Bool, ref: String? = nil, previousValue: MetadataValue? = nil, message: String? = nil) {
        self.applied = applied
        self.ref = ref
        self.previousValue = previousValue
        self.message = message
    }
}
